In [ ]:
products = [
    {"name":"키보드","price":45000, "stock":12, "category":"전자기기"},
    {"name":"모니터","price":45000, "stock":0, "category":"전자기기"},
    {"name":"마우스","price":45000, "stock":6, "category":"전자기기"},
    {"name":"책상","price":45000, "stock":0, "category":"가구"},
    {"name":"의자","price":45000, "stock":8, "category":"가구"},
]{"name":"키보드","price":45000, "stock":5, "category":"전자기기"},

products

[{'name': '키보드', 'price': 45000, 'stock': 12, 'category': '전자기기'},
 {'name': '모니터', 'price': 45000, 'stock': 0, 'category': '전자기기'},
 {'name': '마우스', 'price': 45000, 'stock': 6, 'category': '전자기기'},
 {'name': '책상', 'price': 45000, 'stock': 0, 'category': '가구'},
 {'name': '의자', 'price': 45000, 'stock': 8, 'category': '가구'}]

In [5]:
import pandas as pd

# 1. 엑셀 파일을 불러와 df 변수에 저장합니다.
df = pd.read_csv('book_bestseller_clean.csv')

df.to_excel('book_bestseller_clean.xlsx',index=False)

# 2. 데이터프레임(df)의 첫 5개 행을 출력합니다.
df.head()

,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설


import pandas as pd  
    데이터를 다루는 도구인 pandas 라이브러리를 불러옵니다. 매번 pandas를 다 적기 귀찮으므로 앞으로는 pd라는 짧은 별명으로 부르겠다고 선언하는 것입니다.  
pd.read_excel('파일명.xlsx')  
    pandas(pd) 라이브러리에 있는 엑셀 읽기 기능(read_excel)을 사용해 파일을 읽어옵니다. 파일이 노트북과 같은 폴더에 있으므로 파일 이름만 정확히 적어주면 됩니다.  
df = ...  
    불러온 표 데이터를 df라는 이름의 DataFrame(데이터프레임) 변수에 담아두는 역할을 합니다.  
df.head()  
    데이터프레임의 맨 위(머리)에서부터 5개 행을 가져와서 화면에 보여주는 명령어입니다. 괄호 안에 숫자를 넣지 않으면 기본적으로 5줄을 보여줍니다. (만약 10줄을 보고 싶다면 df.head(10)이라고 쓰면 됩니다.)

In [6]:
print("1.행,열:",df.shape)
print("2.컴럼:",df.columns)
print("3. 각 컬럼의 자료형:",df.dtypes)
print("4.결측치:",df.isna().sum())
print(df.head)

1.행,열: (199, 8)
2.컴럼: Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')
3. 각 컬럼의 자료형: 순위        int64
판매상품ID      str
상품명         str
판매가       int64
저자          str
출판사         str
발행일         str
분야          str
dtype: object
4.결측치: 순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64
<bound method NDFrame.head of       순위         판매상품ID                             상품명    판매가          저자  \
0      1  S000000610612                          소년이 온다  13500          한강   
1      2  S000001632467                              모순  11700         양귀자   
2      3  S000216262757                      결국 국민이 합니다  19800         이재명   
3      4  S000216111714                             혼모노  16200         성해나   
4      5  S000200550189                              급류  12600         정대건   
..   ...            ...                             ...    ...         ...   
194  195  S000001461135                       용의자 X의 

df.shape  
데이터프레임의 (행 개수, 열 개수)를 튜플 형태로 알려줍니다. 괄호 ()를 붙이지 않는 속성입니다.  

df.columns  
표의 맨 위에 있는 모든 컬럼(열) 이름들을 모아서 보여줍니다.  

df.dtypes  
각 컬럼에 들어있는 데이터가 숫자(int, float)인지, 문자열(object)인지 등의 자료형을 보여줍니다.

df.isna().sum()  
df.isna()는 데이터가 비어있으면(결측치) True, 채워져 있으면 False로 표시합니다. 여기에 .sum()을 붙이면 각 컬럼별로 비어있는 값이 몇 개나 되는지 더해서 보여줍니다.

df.head()  
이전 단계에서 보았던 것처럼 데이터의 상위 5개 행을 시각적으로 깔끔하게 출력해 줍니다. (주피터 노트북 환경에서는 한 셀의 가장 마지막 줄에 df.head()를 적어야 표 형태로 예쁘게 출력됩니다.)

df.info()
위에서 확인한 2번(컬럼 이름), 3번(자료형), 4번(결측치 제외 개수)을 한눈에 요약해서 보여주는 df.info()

In [8]:
# 1. 원하는 컬럼 이름들을 리스트로 만듭니다.
columns_to_keep = [
    '순위', 
    '판매상품ID', 
    '상품명', 
    '판매가', 
    '저자', 
    '출판사', 
    '발행일', 
    '분야'
    ]

# 2. 리스트를 사용해 원하는 컬럼만 쏙 골라내어 df_books에 저장합니다.
df_books = df[columns_to_keep]

# 3. 컬럼이 잘 선택되었는지 데이터의 앞 5행을 확인합니다.
df_books.head()

,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설


In [9]:
df_books = df_books.rename(columns={
    '판매상품ID': '발행번호',
})

print("변경 후 컬럼 이름 목록:", df_books.columns)

변경 후 컬럼 이름 목록: Index(['순위', '발행번호', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')


columns={'기존이름': '바꿀이름'}

딕셔너리 형태로 '바꾸고 싶은 원래 컬럼명': '새로 지정할 컬럼명'을 짝지어 전달합니다. 바꾸고 싶은 컬럼이 여러 개라면 위 코드처럼 쉼표(,)로 구분해서 한 번에 적어주면 됩니다. 여기에 적지 않은 나머지 컬럼들은 이름이 그대로 유지됩니다.

df_books = df_books.rename(...)

rename() 함수는 원본 데이터를 바로 바꾸지 않고, 이름이 바뀐 새로운 데이터프레임을 복사해서 반환합니다. 따라서 변경된 결과를 유지하려면 앞에 df_books = 을 붙여서 다시 저장해 주어야 합니다.

초보자 팁: 다시 저장하는 과정이 귀찮다면 df_books.rename(columns={...}, inplace=True) 처럼 inplace=True라는 옵션을 더해주면 원본 데이터가 즉시 수정됩니다. 다만 데이터 분석을 할 때는 안전을 위해 위 코드처럼 다시 대입하는 방식을 더 권장합니다.

In [10]:
# 1. 모든 컬럼의 결측치 개수 확인하기
print("--- [모든 컬럼] 결측치 개수 ---")
print(df_books.isna().sum())

# 2. 결측치가 있는 컬럼만 골라서 보기
print("\n--- [결측치가 있는 컬럼만] 확인 ---")
all_null_counts = df_books.isna().sum()
print(all_null_counts[all_null_counts > 0])

--- [모든 컬럼] 결측치 개수 ---
순위      0
발행번호    0
상품명     0
판매가     0
저자      0
출판사     0
발행일     0
분야      0
dtype: int64

--- [결측치가 있는 컬럼만] 확인 ---
Series([], dtype: int64)


df_books.isna().sum()

isna()가 데이터프레임의 모든 칸을 검사하여 비어있으면 True(1), 채워져 있으면 False(0)로 바꿉니다.

뒤에 .sum()을 붙이면 세로 방향으로 이 숫자들을 모두 더해주므로, 결과적으로 각 컬럼별 결측치의 총개수가 계산됩니다.


all_null_counts[all_null_counts > 0]

위에서 계산한 결측치 개수(all_null_counts) 중에서 비어있는 값이 1개 이상(> 0)인 컬럼만 쏙 골라내어 출력하는 대괄호 [ ] 필터링 문법입니다. 결측치가 0개인 깨끗한 컬럼들은 제외되므로 데이터 구조를 파악할 때 매우 편리합니다.

In [3]:

import pandas as pd

DATA_PATH = "book_bestseller_clean.csv"
df_books = pd.read_csv(DATA_PATH, encoding="utf-8-sig")

print("데이터 크기:", df_books.shape)
print("컬럼:", df_books.columns.tolist())
print("상품명 결측치:", df_books["상품명"].isna().sum())
df_books[["상품명"]].head(10)

데이터 크기: (199, 8)
컬럼: ['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야']
상품명 결측치: 0


,상품명
0,소년이 온다
1,모순
2,결국 국민이 합니다
3,혼모노
4,급류
5,초역 부처의 말
6,청춘의 독서(특별증보판)
7,어른의 행복은 조용하다
8,채식주의자
9,단 한 번의 삶(강물에디션 활판인쇄 한정판)


In [4]:
titles = (
    df_books["상품명"]
    .fillna("")
    .astype(str)
    .str.strip()
)

titles = titles[titles != ""].reset_index(drop=True)
print("사용할 제목 수:", len(titles))
titles.head(10)

사용할 제목 수: 199


0                      소년이 온다
1                          모순
2                  결국 국민이 합니다
3                         혼모노
4                          급류
5                    초역 부처의 말
6               청춘의 독서(특별증보판)
7                어른의 행복은 조용하다
8                       채식주의자
9    단 한 번의 삶(강물에디션 활판인쇄 한정판)
Name: 상품명, dtype: str

In [5]:
sample_docs = [
    "파이썬 데이터 분석",
    "파이썬 머신러닝",
    "데이터 분석 입문",
]

In [8]:

from sklearn.feature_extraction.text import CountVectorizer

count_vectorizer = CountVectorizer()
X_count_sample = count_vectorizer.fit_transform(sample_docs)

print(count_vectorizer)

CountVectorizer()


In [9]:

feature_names = count_vectorizer.get_feature_names_out()

print(feature_names)

['데이터' '머신러닝' '분석' '입문' '파이썬']


In [10]:

X_count_sample.toarray()

array([[1, 0, 1, 0, 1],
       [0, 1, 0, 0, 1],
       [1, 0, 1, 1, 0]])

In [11]:

sample_count_df = pd.DataFrame(
    X_count_sample.toarray(),
    columns=feature_names,
    index=sample_docs,
)

sample_count_df

,데이터,머신러닝,분석,입문,파이썬
파이썬 데이터 분석,1,0,1,0,1
파이썬 머신러닝,0,1,0,0,1
데이터 분석 입문,1,0,1,1,0


In [12]:
repeat_docs = [
    "파이썬 파이썬 데이터",
    "데이터 분석",
]

repeat_vectorizer = CountVectorizer()
X_repeat = repeat_vectorizer.fit_transform(repeat_docs)

repeat_df = pd.DataFrame(
    X_repeat.toarray(),
    columns=repeat_vectorizer.get_feature_names_out(),
    index=repeat_docs,
)

repeat_df

,데이터,분석,파이썬
파이썬 파이썬 데이터,1,0,2
데이터 분석,1,1,0


In [13]:
test_docs = [
    "AI 데이터 분석 R 파이썬",
]

test_vectorizer = CountVectorizer()
X_test = test_vectorizer.fit_transform(test_docs)
print(test_vectorizer.get_feature_names_out())

['ai' '데이터' '분석' '파이썬']


'R'이 빠진 이유는 CountVectorizer의 기본 설정(Default) 때문입니다.  
scikit-learn의 CountVectorizer는 기본적으로 글자 수가 2개 이상인 단어(토큰)만 추출하도록 설계되어 있습니다.   
'R'은 한 글자(1글자) 단어이기 때문에 분석 대상에서 자동으로 제외된 것  

### token_pattern 옵션을 추가하여 1글자 단어도 포함하도록 설정합니다.
test_vectorizer = CountVectorizer(token_pattern=r"(?u)\b\w+\b")

In [15]:

count_vectorizer = CountVectorizer()
X_count = count_vectorizer.fit_transform(titles)
count_terms = count_vectorizer.get_feature_names_out()


print("문서 수:", X_count.shape[0])
print("단어 수:", X_count.shape[1])
print("행렬 크기:", X_count.shape)

문서 수: 199
단어 수: 536
행렬 크기: (199, 536)



X_count.shape[0]
= 사용한 도서 제목 수 = 199

X_count.shape[1]
= CountVectorizer가 만든 단어 수 = 536

In [17]:
list(count_vectorizer.vocabulary_.items())[:20]

[('소년이', 278),
 ('온다', 355),
 ('모순', 194),
 ('결국', 67),
 ('국민이', 84),
 ('합니다', 511),
 ('혼모노', 524),
 ('급류', 92),
 ('초역', 460),
 ('부처의', 236),
 ('청춘의', 458),
 ('독서', 150),
 ('특별증보판', 485),
 ('어른의', 332),
 ('행복은', 517),
 ('조용하다', 432),
 ('채식주의자', 454),
 ('번의', 220),
 ('강물에디션', 57),
 ('활판인쇄', 528)]

list() = 최종 결과물은 리스트 값으로 반환
count_vectorizer.vocabulary_ = 딕셔너리로 되어 있는 값을
.items() = 딕셔너리의 키와 벨류를 한 쌍으로 묶어서 튜플구조로 반환 하는 것.
.vocabulary = count_vecterizer가 데이터를 입력받아 학습을 마친 후, 결과를 내부에 저장해 둔 단어 사전 데이터(딕셔너리)

In [18]:
type(X_count)

scipy.sparse._csr.csr_matrix

In [19]:

X_count_sample.toarray()

array([[1, 0, 1, 0, 1],
       [0, 1, 0, 0, 1],
       [1, 0, 1, 1, 0]])

In [20]:
X_count.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(199, 536))

In [21]:
print(titles.iloc[0])

소년이 온다


In [22]:

first_row = X_count.getrow(0)
indices = first_row.indices
values = first_row.data

first_title_terms = [
    (count_terms[index], value)
    for index, value in zip(indices, values)
]

first_title_terms

[('소년이', np.int64(1)), ('온다', np.int64(1))]

In [23]:

import numpy as np

count_sums = np.asarray(X_count.sum(axis=0)).ravel()

count_summary = pd.DataFrame({
    "단어": count_terms,
    "전체등장횟수": count_sums,
})

count_summary = count_summary.sort_values(
    "전체등장횟수",
    ascending=False,
).reset_index(drop=True)

count_summary.head(30)

,단어,전체등장횟수
0,에디션,12
1,리커버,8
2,기념,8
3,위한,7
4,해커스,7
5,토익,6
6,나는,5
7,내가,5
8,스페셜,5
9,싶은,4


In [24]:
count_top30 = count_summary.head(30)

count_top30.to_csv(
    "chapter03_count_top_terms.csv",
    index=False,
    encoding="utf-8-sig",
)

In [25]:

pd.read_csv(
    "chapter03_count_top_terms.csv",
    encoding="utf-8-sig",
).head()

,단어,전체등장횟수
0,에디션,12
1,리커버,8
2,기념,8
3,위한,7
4,해커스,7


In [ ]:
# import pandas as pd
# from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# # 1. 예시 문서 정의 (모든 문서에 '책', '도서'가 공통으로 들어감)
# docs = [
#     "파이썬 데이터 분석 책 추천",
#     "딥러닝 인공지능 도서 추천",
#     "소설 에세이 베스트셀러 책 도서 리뷰"
# ]

# # 2. Count 방식 (CountVectorizer)
# count_vec = CountVectorizer()
# count_matrix = count_vec.fit_transform(docs)
# words = count_vec.get_feature_names_out()

# df_count = pd.DataFrame(count_matrix.toarray(), columns=words)
# print("=== [1] CountVectorizer 결과 ===")
# print(df_count[['책', '도서', '파이썬', '딥러닝']])
# print("-" * 50)

# # 3. TF-IDF 방식 (TfidfVectorizer)
# tfidf_vec = TfidfVectorizer()
# tfidf_matrix = tfidf_vec.fit_transform(docs)

# df_tfidf = pd.DataFrame(tfidf_matrix.toarray(), columns=words)
# print("=== [2] TfidfVectorizer 결과 ===")

# === 변경 전 (오류 발생 코드) ===
# print(df_count[['책', '도서', '파이썬', '딥러닝']])

# === 변경 후 (이렇게 수정해 보세요!) ===
# 1. CountVectorizer 결과 전체 출력
# print(df_count) 

# print("-" * 50)

# # 2. TfidfVectorizer 결과 전체 출력 (아래쪽 코드도 함께 수정)
# print("=== [2] TfidfVectorizer 결과 ===")
# print(df_tfidf.round(2))
# # 코드를 사용할 때는 주의가 필요합니다.


=== [1] CountVectorizer 결과 ===
   데이터  도서  딥러닝  리뷰  베스트셀러  분석  소설  에세이  인공지능  추천  파이썬
0    1   0    0   0      0   1   0    0     0   1    1
1    0   1    1   0      0   0   0    0     1   1    0
2    0   1    0   1      1   0   1    1     0   0    0
--------------------------------------------------
=== [2] TfidfVectorizer 결과 ===


NameError: name 'df_tfidf' is not defined

In [30]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# 1. 예시 문서 정의
docs = [
    "파이썬 데이터 분석 책 추천",
    "딥러닝 인공지능 도서 추천",
    "소설 에세이 베스트셀러 책 도서 리뷰"
]

# 2. Count 방식
count_vec = CountVectorizer(token_pattern=r"(?u)\b\w+\b")
count_matrix = count_vec.fit_transform(docs)
words = count_vec.get_feature_names_out()

df_count = pd.DataFrame(count_matrix.toarray(), columns=words)
print("=== [1] CountVectorizer 결과 ===")
print(df_count)

print("-" * 50)

# 3. TF-IDF 방식
tfidf_vec = TfidfVectorizer(token_pattern=r"(?u)\b\w+\b")
tfidf_matrix = tfidf_vec.fit_transform(docs)

# 이 부분의 주석(#)이 완전히 해제되어 있어야 에러가 나지 않습니다.
df_tfidf = pd.DataFrame(tfidf_matrix.toarray(), columns=words)
print("=== [2] TfidfVectorizer 결과 ===")
print(df_tfidf.round(2))

=== [1] CountVectorizer 결과 ===
   데이터  도서  딥러닝  리뷰  베스트셀러  분석  소설  에세이  인공지능  책  추천  파이썬
0    1   0    0   0      0   1   0    0     0  1   1    1
1    0   1    1   0      0   0   0    0     1  0   1    0
2    0   1    0   1      1   0   1    1     0  1   0    0
--------------------------------------------------
=== [2] TfidfVectorizer 결과 ===
    데이터    도서   딥러닝    리뷰  베스트셀러    분석    소설   에세이  인공지능     책    추천   파이썬
0  0.49  0.00  0.00  0.00   0.00  0.49  0.00  0.00  0.00  0.37  0.37  0.49
1  0.00  0.43  0.56  0.00   0.00  0.00  0.00  0.00  0.56  0.00  0.43  0.00
2  0.00  0.33  0.00  0.44   0.44  0.00  0.44  0.44  0.00  0.33  0.00  0.00


In [31]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# 1. 이미지에 나온 예시 문서 정의 (파이썬 2번, 데이터 1번, 분석 1번)
docs = [
    "파이썬 파이썬 데이터 분석"
]

# 2. 순수한 등장 횟수(TF) 확인을 위한 CountVectorizer
count_vec = CountVectorizer()
count_matrix = count_vec.fit_transform(docs)
words = count_vec.get_feature_names_out()

df_count = pd.DataFrame(count_matrix.toarray(), columns=words)
print("=== [1] 단순 등장 횟수 (Count) ===")
print(df_count)
print("-" * 50)

# 3. scikit-learn의 정규화가 적용된 TfidfVectorizer
tfidf_vec = TfidfVectorizer()
tfidf_matrix = tfidf_vec.fit_transform(docs)

df_tfidf = pd.DataFrame(tfidf_matrix.toarray(), columns=words)
print("=== [2] scikit-learn의 최종 TF-IDF 값 ===")
print(df_tfidf.round(2))


=== [1] 단순 등장 횟수 (Count) ===
   데이터  분석  파이썬
0    1   1    2
--------------------------------------------------
=== [2] scikit-learn의 최종 TF-IDF 값 ===
    데이터    분석   파이썬
0  0.41  0.41  0.82
